pyteal/ast/router.py

```diff
@@ -0,0 +1,583 @@
+                    OutputKwArgInfo, handler.output_kwarg_info
+                ).abi_type.new_instance()
+                subroutine_call: abi.ReturnedValue = cast(
+                    abi.ReturnedValue, handler(*arg_abi_vars)
+                )
+                return Seq(
+                    *decode_instructions,
+                    subroutine_call.store_into(output_temp),
+                    abi.MethodReturn(output_temp),
+                    Approve(),
+                )
+
+    def __add_bare_call(self, oc_actions: OnCompleteActions) -> None:
+        action_type = Expr | SubroutineFnWrapper | ABIReturnSubroutine
+        if oc_actions.is_empty():
+            raise TealInternalError("the OnCompleteActions is empty.")
+        cs_calls, approval_calls = oc_actions._clear_state_n_other_oc()
+        if cs_calls.on_call:
+            on_call = cast(action_type, cs_calls.on_call)
+            wrapped = Router._wrap_handler(False, on_call)
+            self.categorized_clear_state_ast.bare_calls.append(
+                CondNode(Int(1), wrapped)
+            )
+        if cs_calls.on_create:
+            on_create = cast(action_type, cs_calls.on_create)
+            wrapped = Router._wrap_handler(False, on_create)
+            self.categorized_clear_state_ast.bare_calls_create.append(
+                CondNode(Txn.application_id() == Int(0), wrapped)
+            )
+        for oc, approval_bac in approval_calls.items():
+            if approval_bac.on_call:
+                on_call = cast(action_type, approval_bac.on_call)
+                wrapped = Router._wrap_handler(False, on_call)
+                self.categorized_approval_ast.bare_calls.append(
+                    CondNode(Txn.on_completion() == oc, wrapped)
+                )
+            if approval_bac.on_create:
+                on_create = cast(action_type, approval_bac.on_create)
+                wrapped = Router._wrap_handler(False, on_create)
+                self.categorized_approval_ast.bare_calls_create.append(
+                    CondNode(
+                        And(Txn.application_id() == Int(0), Txn.on_completion() == oc),
+                        wrapped,
+                    )
+                )
+
+    def add_method_handler(
+        self,
+        method_call: ABIReturnSubroutine,
+        overriding_name: str = None,
+        call_configs: CallConfigs = CallConfigs(),
+    ) -> None:
+        if not isinstance(method_call, ABIReturnSubroutine):
+            raise TealInputError(
+                "for adding method handler, must be ABIReturnSubroutine"
+            )
+        method_signature = method_call.method_signature(overriding_name)
+        if call_configs.is_never():
+            raise TealInputError(
+                f"registered method {method_signature} is never executed"
+            )
+        if method_signature in self.added_method_sig:
+            raise TealInputError(f"re-registering method {method_signature} detected")
```

Comment:
Note that there are two levels of collisions we must avoid:
* Registering the exact same method signature twice (covered here).
* Registering a method whose selector (8 byte hash of signature) collides with another method (not covered as far as I can tell).

Reply:
👍